src/pint/models/model_sector.py

```diff
@@ -0,0 +1,510 @@
+""" Defining Model Sector, classes for grouping the same type of model components and provide
+the uniformed API 
+"""
+import numpy as np
+import astropy.units as u
+from astropy import time
+import copy
+from collections import defaultdict
+
+from pint.utils import get_component_type
+from pint.phase import Phase
+
+
+class ModelSector(object):
+    """ A class that groups the same type of component and provide the API for
+    gathering the information from each component.
+
+    Parameter
+    ---------
+    components: list of `Component` sub-class object.
+        Components that are belong to the same type, for example, DelayComponent.
+
+    Note
+    ----
+    The order of the component in the list is the order a component get computed.
+    """
+
+    _methods = tuple()
```

Comment:
What is `_methods` supposed to contain? Presumably a list of some subset of the methods provided by objects in this sector? Not all methods, one assumes? A name that described which methods should be listed would be useful.

It also looks like this has to be populated manually, by listing the methods - it would be preferable to create a decorator one could attach to a method that wanted to be listed in its class' _methods list.

If order doesn't matter, a set or a frozenset would be appropriate here, especially if you're using intersection operators on it.